Parcelable auto value entities are being generated incorrectly with the 1.0.0 version

When updating auto value moshi dependencies from `0.4.7` to `1.0.0` the auto value entities which are marked as `Parcelable` started failing (does not compile). Tests are done using the latest version (`0.2.8`) of `auto-parcel`. It seems that new version (`1.0.0`) is adding redundant `$` sign while calling a constructor in one of the generated classes.

Here is a simple auto value class:
```java
@AutoValue
public abstract class Foo implements Parcelable {
    abstract String bar();

    public static JsonAdapter<Foo> jsonAdapter(Moshi moshi) {
        return new AutoValue_Foo.MoshiJsonAdapter(moshi);
    }
}
```

Here is the generated `$AutoValue_Foo` class: 
A comment is added to the line that has the wrong constructor call. It seems that `$AutoValue_Foo(bar)` should be `AutoValue_Foo(bar)`.
```java
abstract class $AutoValue_Foo extends $$AutoValue_Foo {
  $AutoValue_Foo(String bar) {
    super(bar);
  }

  public static final class MoshiJsonAdapter extends JsonAdapter<Foo> {
    private static final String[] NAMES = new String[] {"bar"};
    private static final JsonReader.Options OPTIONS = JsonReader.Options.of(NAMES);
    private final JsonAdapter<String> barAdapter;
    public MoshiJsonAdapter(Moshi moshi) {
      this.barAdapter = moshi.adapter(String.class).nonNull();
    }
    @Override
    public Foo fromJson(JsonReader reader) throws IOException {
      reader.beginObject();
      String bar = null;
      while (reader.hasNext()) {
        switch (reader.selectName(OPTIONS)) {
          case 0: {
            bar = this.barAdapter.fromJson(reader);
            break;
          }
          case -1: {
            // Unknown name, skip it
            reader.skipName();
            reader.skipValue();
          }
        }
      }
      reader.endObject();
      return new $AutoValue_Foo(bar); // $AutoValue_Foo(bar) -> AutoValue_Foo(bar)
    }
    @Override
    public void toJson(JsonWriter writer, Foo value) throws IOException {
      writer.beginObject();
      writer.name("bar");
      this.barAdapter.toJson(writer, value.bar());
      writer.endObject();
    }
    @Override
    public String toString() {
      return new StringBuilder().append("JsonAdapter(").append("Foo").append(")").toString();
    }
  }
}
```

Here is the generated code with old version (`0.4.7`) using the same auto value class definition.
```java
abstract class $AutoValue_Foo extends $$AutoValue_Foo {
  $AutoValue_Foo(String bar) {
    super(bar);
  }

  public static final class MoshiJsonAdapter extends JsonAdapter<Foo> {
    private static final String[] NAMES = new String[] {"bar"};
    private static final JsonReader.Options OPTIONS = JsonReader.Options.of(NAMES);
    private final JsonAdapter<String> barAdapter;
    public MoshiJsonAdapter(Moshi moshi) {
      this.barAdapter = adapter(moshi, String.class);
    }
    @Override
    public Foo fromJson(JsonReader reader) throws IOException {
      reader.beginObject();
      String bar = null;
      while (reader.hasNext()) {
        switch (reader.selectName(OPTIONS)) {
          case 0: {
            bar = this.barAdapter.fromJson(reader);
            break;
          }
          case -1: {
            // Unknown name, skip it
            reader.skipName();
            reader.skipValue();
          }
        }
      }
      reader.endObject();
      return new AutoValue_Foo(bar); // This is correct constructor call
    }
    @Override
    public void toJson(JsonWriter writer, Foo value) throws IOException {
      writer.beginObject();
      writer.name("bar");
      this.barAdapter.toJson(writer, value.bar());
      writer.endObject();
    }
    private JsonAdapter adapter(Moshi moshi, Type adapterType) {
      return moshi.adapter(adapterType);
    }
  }
}
```

@ZacSweers I understand that this in not the perfect solution. This was the working solution in `0.4.7` and I have just copied it to here. Here is the link to the working solution. 

https://github.com/rharter/auto-value-moshi/blob/ef6bcd70733e3570c7ee070d30bdea918003ce87/auto-value-moshi/src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtension.java#L548

Since you have better know-how in `auto-value-moshi` and `auto-value_parcel` it would be better you propose a better solution. But it is kinda obvious that when `AutoValueMoshiExtension` class is rewritten it is broken because the outputs are not same any more. You can check again the outputs  of `0.4.7` and `1.0.0` from the first comment. 

I think `auto-value-moshi` should not make an assumption about single hierarchy. There might be other abstract classes in the hierarchy while generating the code as in the case of `Parcelable`s. 

Here is the full error code (Compile error) when I try to generate code for the Foo class in first comment.

```
.../entity/$AutoValue_Foo.java:42: error: $AutoValue_Foo is abstract; cannot be instantiated
      return new $AutoValue_Foo(bar);
             ^
``` 

Here is the generated class:

```java
package de.meinfernbus.network.entity;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.JsonReader;
import com.squareup.moshi.JsonWriter;
import com.squareup.moshi.Moshi;
import java.io.IOException;
import java.lang.Override;
import java.lang.String;
import java.lang.StringBuilder;

abstract class $AutoValue_Foo extends $$AutoValue_Foo {
  $AutoValue_Foo(String bar) {
    super(bar);
  }

  public static final class MoshiJsonAdapter extends JsonAdapter<Foo> {
    private static final String[] NAMES = new String[] {"bar"};
    private static final JsonReader.Options OPTIONS = JsonReader.Options.of(NAMES);
    private final JsonAdapter<String> barAdapter;
    public MoshiJsonAdapter(Moshi moshi) {
      this.barAdapter = moshi.adapter(String.class).nonNull();
    }
    @Override
    public Foo fromJson(JsonReader reader) throws IOException {
      reader.beginObject();
      String bar = null;
      while (reader.hasNext()) {
        switch (reader.selectName(OPTIONS)) {
          case 0: {
            bar = this.barAdapter.fromJson(reader);
            break;
          }
          case -1: {
            // Unknown name, skip it
            reader.skipName();
            reader.skipValue();
          }
        }
      }
      reader.endObject();
      return new $AutoValue_Foo(bar);
    }
    @Override
    public void toJson(JsonWriter writer, Foo value) throws IOException {
      writer.beginObject();
      writer.name("bar");
      this.barAdapter.toJson(writer, value.bar());
      writer.endObject();
    }
    @Override
    public String toString() {
      return new StringBuilder().append("JsonAdapter(").append("Foo").append(")").toString();
    }
  }
}

```

It should not call `$AutoValue_Foo(bar)` but `AutoValue_Foo(bar)` in return statement of `fromJson` method.

--- a/auto-value-moshi-extension/src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtension.java
+++ b/auto-value-moshi-extension/src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtension.java
@@ -243,8 +243,7 @@ public final class AutoValueMoshiExtension extends AutoValueExtension {
         autoValueClassName,
         genericTypeNames,
         properties,
-        context.builder().orElse(null),
-        context.processingEnvironment(),
+        context,
         adapterClassName);
 
     Optional<AnnotationSpec> generatedAnnotation = GeneratedAnnotationSpecs.generatedAnnotationSpec(
@@ -419,8 +418,7 @@ public final class AutoValueMoshiExtension extends AutoValueExtension {
       ClassName autoValueClassName,
       TypeVariableName[] genericTypeNames,
       List<Property> properties,
-      @Nullable BuilderContext builderContext,
-      ProcessingEnvironment processingEnvironment,
+      Context context,
       String adapterClassName
   ) {
 
@@ -504,7 +502,7 @@ public final class AutoValueMoshiExtension extends AutoValueExtension {
         .addFields(adapters.values())
         .addMethod(constructor.build())
         .addMethod(createReadMethod(className, autoValueClassName, autoValueTypeName, properties,
-                adapters, names, builderContext, processingEnvironment))
+                adapters, names, context))
         .addMethod(createWriteMethod(autoValueTypeName, adapters))
         .addMethod(MethodSpec.methodBuilder("toString")
             .addAnnotation(Override.class)
@@ -591,8 +589,7 @@ public final class AutoValueMoshiExtension extends AutoValueExtension {
 
   private MethodSpec createReadMethod(ClassName className, ClassName autoValueClassName, TypeName autoValueTypeName,
                                       List<Property> properties, ImmutableMap<Property, FieldSpec> adapters,
-                                      List<String> names, @Nullable BuilderContext builderContext,
-                                      ProcessingEnvironment processingEnvironment) {
+                                      List<String> names, Context context) {
     NameAllocator nameAllocator = new NameAllocator();
     ParameterSpec reader = ParameterSpec.builder(JsonReader.class, nameAllocator.newName("reader"))
         .build();
@@ -602,10 +599,11 @@ public final class AutoValueMoshiExtension extends AutoValueExtension {
         .returns(autoValueTypeName)
         .addParameter(reader)
         .addException(IOException.class);
+    BuilderContext builderContext = context.builder().orElse(null);
     // Validate the builderContext if there is one.
     if (builderContext != null) {
       if (!builderContext.buildMethod().isPresent()) {
-        processingEnvironment.getMessager()
+        context.processingEnvironment().getMessager()
                 .printMessage(
                         Diagnostic.Kind.ERROR,
                         "Could not determine the build method. Make sure it is named \"build\".",
@@ -621,7 +619,7 @@ public final class AutoValueMoshiExtension extends AutoValueExtension {
                 .collect(Collectors.toSet());
 
         if (annotatedMethods.size() > 1) {
-          processingEnvironment.getMessager()
+          context.processingEnvironment().getMessager()
                   .printMessage(
                           Diagnostic.Kind.ERROR,
                           "Too many @AutoValueMoshiBuilder annotated builder methods.",
@@ -631,7 +629,7 @@ public final class AutoValueMoshiExtension extends AutoValueExtension {
         }
 
         if (annotatedMethods.isEmpty()) {
-          processingEnvironment.getMessager().printMessage(
+          context.processingEnvironment().getMessager().printMessage(
                   Diagnostic.Kind.ERROR,
                   "Too many builder methods. Annotate builder method with @AutoValueMoshiBuilder.",
                   builderMethods.stream().findAny().get()
@@ -730,7 +728,8 @@ public final class AutoValueMoshiExtension extends AutoValueExtension {
       readMethod.addStatement("return $N.$L", builderField.get(), builderContext.buildMethod().get());
     } else {
       CodeBlock params = CodeBlock.join(constructorCall, ", ");
-      readMethod.addStatement("return new $T($L)", className, params);
+      ClassName constructorName = ClassName.bestGuess(context.finalAutoValueClassName());
+      readMethod.addStatement("return new $T($L)", constructorName, params);
     }
     return readMethod.build();
   }

